File: codeforces/1900~1999/1955/h.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

constexpr int MAX_R = 13;

void solve() {
  int n = 0, m = 0, k = 0;
  cin >> n >> m >> k;

  vector<string> grid(n);
  for (int i = 0; i < n; ++i) cin >> grid[i];

  vector<int> x(k), y(k), p(k);
  for (int i = 0; i < k; ++i) {
    cin >> x[i] >> y[i] >> p[i];
    x[i]--;
    y[i]--;
  }

  vector<i64> p3 = {1};
  for (int i = 1; i < MAX_R; ++i) {
    p3.push_back(p3.back() * 3);
  }

  auto compute_damage = [&](int r, int c, int d, int len) -> i64 {
    int health = 0;
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < m; ++j) {
        if (grid[i][j] == '#') {
          int dx = r-i, dy = c-j;
          if (dx * dx + dy * dy <= len * len) {
            health += d;
          }
        }
      }
    }
    return health;
  };

  vector<vector<i64>> damages(k, vector<i64>(MAX_R));
  for (int i = 0; i < k; ++i) {
    for (int rad = 0; rad < MAX_R; ++rad) {
      damages[i][rad] = compute_damage(x[i], y[i], p[i], rad);
    }
  }

  int mask_size = 1 << MAX_R;
  vector<vector<i64>> dp(k+1, vector<i64>(mask_size));
  for (int i = 1; i <= k; ++i) {
    for (int mask = 0; mask < mask_size; ++mask) {
      dp[i][mask] = dp[i-1][mask];
      for (int rad = 0; rad < MAX_R; ++rad) if (mask & (1<<rad)) {
        dp[i][mask] = max(dp[i][mask], dp[i-1][mask ^ (1<<rad)] + damages[i-1][rad]);
      }
    }
  }

  i64 ans = 0;
  for (int mask = 1; mask < mask_size; ++mask) {
    i64 power3 = 0;
    for (int rad = 0; rad < MAX_R; ++rad) if (mask & (1<<rad)) {
      power3 += p3[rad];
    }
    ans = max(ans, dp[k][mask] - power3);
  }

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1955/h.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn read_str() -> String {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

const MAX_R: usize = 12;

fn compute_damage(x: i32, y: i32, p: i32, r: i32, grid: &Vec<Vec<char>>) -> i32 {
    let n = grid.len();
    let m = grid[0].len();
    let mut ans = 0;
    for i in 0..n {
        for j in 0..m {
            if grid[i][j] == '#' {
                let dx = i as i32 - x;
                let dy = j as i32 - y;
                let s = dx * dx + dy * dy;
                if s <= r * r {
                    ans += p;
                }
            }
        }
    }
    ans
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let inputs = read_vec::<usize>();
    let n = inputs[0];
    let k = inputs[2];

    let mut grid: Vec<Vec<char>> = Vec::new();
    for _i in 0..n {
        grid.push(read_str().chars().collect());
    }

    let mut x: Vec<i32> = vec![0; k+1];
    let mut y: Vec<i32> = vec![0; k+1];
    let mut p: Vec<i32> = vec![0; k+1];
    for i in 1..=k {
        let entry = read_vec::<i32>();
        x[i] = entry[0] - 1;
        y[i] = entry[1] - 1;
        p[i] = entry[2];
    }

    let mut damage = vec![vec![0; MAX_R+1]; k+1];
    for i in 1..=k {
        for r in 0..=MAX_R {
            damage[i][r] = compute_damage(x[i], y[i], p[i], r as i32, &grid);
        }
    }

    let mask_size = 1 << (MAX_R + 1);
    let mut dp = vec![vec![0; mask_size]; k+1];
    for i in 1..=k {
        for mask in 0..mask_size {
            dp[i][mask] = dp[i][mask].max(dp[i-1][mask]);
            for r in 0..=MAX_R {
                if mask & (1 << r) > 0 {
                    let old_mask = mask ^ (1 << r);
                    dp[i][mask] = dp[i][mask].max(dp[i-1][old_mask] + damage[i][r]);
                }
            }
        }
    }

    let mut p3: Vec<i64> = vec![1];
    for i in 1..=MAX_R {
        p3.push(p3[i-1] * 3);
    }

    let mut ans = 0;
    for mask in 0..mask_size {
        let mut p3_sum = 0;
        for r in 0..=MAX_R {
            if mask & (1 << r) > 0 {
                p3_sum += p3[r];
            }
        }
        ans = ans.max(dp[k][mask] as i64 - p3_sum);
    }

    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
